### PR TITLE
introduce bin-version crate for defining binary version strings

### PR DESCRIPTION
**Pull request by bmwill**
## Description 

This dedups the binary version code we have copy-pasted in various different places in our codebase and centralizes the logic in the bin-version crate. This crate defines macros which defines the `GIT_REVISION` and `VERSION` consts as well as adds compile-time checks to ensure these are only defined in binaries and not from libraries (which could result in invalid git revision values).

## Test plan 

How did you test the new or updated feature?

---

## Release notes

Check each box that your changes affect. If none of the boxes relate to your changes, release notes aren't required.

For each box you select, include information after the relevant heading that describes the impact of your changes that a user might notice and any actions they must take to implement updates. 

- [ ] Protocol: 
- [ ] Nodes (Validators and Full nodes): 
- [ ] Indexer: 
- [ ] JSON-RPC: 
- [ ] GraphQL: 
- [ ] CLI: 
- [ ] Rust SDK:

### CODE REVIEW

**Comment by bmwill** [comment]
@chris-gorham the only place i didn't change was the one in sui-tool that needs to be moved into a binary before it can be converted to use this new crate/macro.

File: crates/sui-graphql-rpc/src/main.rs [hunk 1]
```diff
@@ -14,20 +14,8 @@ use sui_graphql_rpc::server::graphiql_server::start_graphiql_server;
...
-        revision
-    } else {
-        git_version::git_version!(
-            args = ["--always", "--abbrev=40", "--dirty", "--exclude", "*"],
-            fallback = "DIRTY"
-        )
-    }
-};
+// Define the `GIT_REVISION` const
+bin_version::git_revision!();
```
Comment (stefan-mysten) [inline crates/sui-graphql-rpc/src/main.rs:18]: Please note that for GraphQL we have set the `abbrev` to 40 as per @ebmifa's request. Can we pass an argument to this `git_revision` to specify the length of the `sha`?

Comment (bmwill) [reply crates/sui-graphql-rpc/src/main.rs:18]: What's the rationale for this vs having it be consistent with what we do elsewhere?

Comment (stefan-mysten) [reply crates/sui-graphql-rpc/src/main.rs:18]: Makes it simpler for deployments. @ebmifa could probably chime in more.

Comment (ebmifa) [reply crates/sui-graphql-rpc/src/main.rs:18]: I don't remember why I wanted a full sha now 🤦🏼 
But, for consistency's sake, let's keep it abbreviated for now.

Comment (stefan-mysten) [reply crates/sui-graphql-rpc/src/main.rs:18]: Ok, then I think somewhere in the GraphQL deployments/CI stuff we need to get the full SHA for the rest to work. You might remember where, so just mentioning this.

**Review by chris-gorham — APPROVED** [review]
LGTM   🔥

**Pull request merged**